nixos/modules/services/web-apps/nextcloud-backup.nix

```diff
@@ -0,0 +1,196 @@
+    NEXTCLOUD_DB_TYPE=$(${lib.getExe pkgs.jq} --raw-output '.databaseType' $METADATA_FILE)
+    # A few sanity checks to make sure that it's possible to restore the backup.
+    if [ ! "$NEXTCLOUD_MAJOR_VERSION" = "${lib.versions.major cfg.package.version}" ] ;then
+      echo "The backup is for Nextcloud $NEXTCLOUD_MAJOR_VERSION but Nextcloud ${lib.versions.major cfg.package.version} is installed."
+      echo "Please change to Nextcloud $NEXTCLOUD_MAJOR_VERSION before restoring (and then possibly upgrade), aborting."
+      exit 1
+    fi
+    if [ ! "$NEXTCLOUD_DB_TYPE" = "${cfg.config.dbtype}" ] ;then
+      echo "The backup is for the database $NEXTCLOUD_DB_TYPE but the current installation uses ${cfg.config.dbtype}, aborting."
+      exit 1
+    fi
+    ${checkBackup}
+    if [ ! -e "${cfg.datadir}/config/override.config.php" ] ;then
+      echo "Expected to find the file ${cfg.datadir}/config/override.config.php but it seems to be missing."
+      echo "It appears that nextcloud has not been set up correctly, aborting."
+      exit 1
+    fi
+    printf "WARNING: Restoring the backup will overwrite the current nextcloud instance. Continue? (y/n):"
+    read answer
+    if [ "$answer" = "''${answer#[Yy]}" ] ;then
+        echo "Restore cancelled, aborting."
+        exit 1
+    fi
+    echo "Enabling maintenance mode..."
+    ${lib.getExe cfg.occ} maintenance:mode --on
+    echo "Restoring config and manually installed apps..."
+    ${restoreRsyncCommand} $1/nextcloud-backup/config.php ${cfg.datadir}/config/
+    ${restoreRsyncCommand} $1/nextcloud-backup/store-apps ${cfg.home}
+    echo "Restoring data directory..."
+    ${restoreRsyncCommand} $1/nextcloud-backup/data ${cfg.datadir}
+    echo "Restoring database..."
+    ${dbBackupRestore.restoreCommand}
+    echo "Disabling maintenance mode..."
+    ${lib.getExe cfg.occ} maintenance:mode --off
+    echo "Nextcloud backup successfully restored."
+    echo "If the backup was not very recent, it might make sense to run \"nextcloud-occ maintenance:data-fingerprint\"."
+    echo "See https://docs.nextcloud.com/server/latest/admin_manual/maintenance/restore.html#synchronising-with-clients-after-data-recovery for more information."
+  '';
+in
+{
+  options.services.nextcloudBackup = {
+    backupScript = lib.mkOption {
+      type = lib.types.package;
+      default = backupScript;
+      defaultText = lib.literalMD "generated script";
+      internal = true;
+      description = lib.mdDoc ''
+        A script that backs up this Nextcloud instance to a given folder.
+      '';
+    };
+    restoreScript = lib.mkOption {
+      type = lib.types.package;
+      default = restoreScript;
+      defaultText = lib.literalMD "generated script";
+      internal = true;
+      description = lib.mdDoc ''
+        A script that restores the Nextcloud instance from a given backup.
+      '';
+    };
+  };
+  config = lib.mkIf cfg.enable {
```

Comment:
I think this should be a separate enable options because not everyone might need these scripts.